libraries/state_history/trace_converter.cpp

```diff
@@ -30,25 +38,262 @@ void trace_converter::add_transaction(const transaction_trace_ptr& trace, const
+}
+
+template <typename Visitor>
+void do_visit_trace(const char* read_buffer, fc::datastream<const char*>& strm, transaction_trace& trace,
+                    compression_type compression, Visitor&& visitor) {
+   auto& trace_v0 = trace.get<transaction_trace_v0>();
+   if (trace_v0.failed_dtrx_trace.size()) {
+      // failed_dtrx_trace have at most one element because it is encoded as an optional
+      do_visit_trace(read_buffer, strm, trace_v0.failed_dtrx_trace[0].recurse, compression, visitor);
+   }
+   if (trace_v0.partial) {
+      prunable_data_type prunable;
+      unpack_prunable(read_buffer, strm, prunable, compression);
+      visitor(trace_v0, prunable);
+   }
+}
+
+struct restore_partial {
+   partial_transaction_v0& ptrx;
+   void operator()(prunable_data_type::full_legacy& data) const {
+      // this should never happen because we convert full_legacy to full before serialization
+      assert(false);
+   }
+   void operator()(prunable_data_type::none& data) const {}
+   void operator()(prunable_data_type::signatures_only& data) const {
+      ptrx.signatures        = std::move(data.signatures);
+   }
+   void operator()(prunable_data_type::partial& data) const  {
+      ptrx.signatures        = std::move(data.signatures);
+   }
+   void operator()(prunable_data_type::full& data) const {
+      ptrx.signatures        = std::move(data.signatures);
+      ptrx.context_free_data = std::move(data.context_free_segments);
+   }
+};
+
+} // namespace
+
+bytes trace_converter::pack(const chainbase::database& db, bool trace_debug_mode, const block_state_ptr& block_state,
+                            uint32_t version, compression_type compression) {
 
-   return fc::raw::pack(make_history_context_wrapper(db, trace_debug_mode, traces));
+   auto traces     = prepare_traces(*this, block_state);
+   auto unprunable = zlib_compress_bytes(fc::raw::pack(make_history_context_wrapper(
+       db, trace_receipt_context{.debug_mode = trace_debug_mode, .version = version}, traces)));
+
+   if (version == 0) {
+      return unprunable;
+   } else {
+      // In version 1 of ShiP traces log disk format, it log entry consists of 3 parts.
+      //  1. a unit8_t compression tag to indicate the compression level of prunables
+      //  2. an zlib compressed unprunable section contains the serialization of the vector of traces excluding
+      //     the prunable_data data (i.e. signatures and context free data)
+      //  3. a prunable section contains the serialization of the vector of optional (possibly compressed)
+      //     prunable_data
+
+      auto                   prunables = pack_prunables(traces, compression);
+      fc::datastream<size_t> unprunable_size_strm;
+      fc::raw::pack(unprunable_size_strm, unprunable);
+      uint8_t compression_tag = static_cast<uint8_t>(compression);
+      bytes   buffer(sizeof(compression_tag) + unprunable_size_strm.tellp() + prunables.size());
```

Comment:
I think you should use `pack` to tell you the pack size of compression_tag instead of using sizeof.

Reply:
code written, no long exists.